Icons pack, implements #303

ilmich opened the pull request:
My first iteration of icons pack support. I've tested with Numix Circle icons pack and Flat icons pack available on play store.
Surely there is some optimization (cache for generated bitmap in case there is no icon in the pack for a particular app), but just for review.

I've also defined the custom intent filter `fr.neamar.kiss.THEMES` for icons pack developers.
Soon some screenshot ;)

Cheers
Michele

ilmich commented:
![screenshot_2015-11-21-13-05-27_redux](https://cloud.githubusercontent.com/assets/148796/11318414/959efe72-9051-11e5-8b48-62f047bb2055.png)
![screenshot_2015-11-21-13-05-21_redux](https://cloud.githubusercontent.com/assets/148796/11318415/99554878-9051-11e5-98f7-e3b5f68bbccb.png)
![screenshot_2015-11-21-13-05-45_redux](https://cloud.githubusercontent.com/assets/148796/11318416/9f6035f2-9051-11e5-88aa-3fbda0075ff1.png)
![screenshot_2015-11-21-13-05-58_redux](https://cloud.githubusercontent.com/assets/148796/11318417/a3670d42-9051-11e5-9a99-f9deac7a923c.png)

ilmich commented:
In this screenshoots F-Droid is a generated icon (not in the icons pack)

ilmich commented:
At this time only Apps are themed, but is easy to implement icons theming for everything (toggle, contacts, shortcuts and so on)

saveman71 commented:
Looks great! Well done man :+1:

saveman71 commented:
I commented on the commits :sob: Will move the comments, sorry if you were spammed...

saveman71 commented on app/src/main/java/fr/neamar/kiss/IconsHandler.java:
```diff
@@ -0,0 +1,283 @@
+            if (xpp != null) {
+                int eventType = xpp.getEventType();
+                while (eventType != XmlPullParser.END_DOCUMENT) {
+                    if(eventType == XmlPullParser.START_TAG) {
```
`if (`
Also applies to for/while loops and others ;)

saveman71 commented on app/src/main/java/fr/neamar/kiss/IconsHandler.java:
```diff
@@ -0,0 +1,283 @@
+                int eventType = xpp.getEventType();
+                while (eventType != XmlPullParser.END_DOCUMENT) {
+                    if(eventType == XmlPullParser.START_TAG) {
+                        if (xpp.getName().equals("iconback"))
```
`if (...) {`
Also applies to for/while loops and others ;)

saveman71 commented on app/src/main/java/fr/neamar/kiss/IconsHandler.java:
```diff
@@ -0,0 +1,283 @@
+                                factor = Float.valueOf(xpp.getAttributeValue(0));
+                            }
+                        }
+                        if (xpp.getName().equals("item")) {
```
`else if`

saveman71 commented on app/src/main/java/fr/neamar/kiss/IconsHandler.java:
```diff
@@ -0,0 +1,283 @@
+                                    drawableName = xpp.getAttributeValue(i);
+                                }
+                            }
+                            if (!packagesDrawables.containsKey(componentName))
```
```
if (...) {
    // ...
}
```
Also applies to for/while loops and others ;)

saveman71 commented on app/src/main/java/fr/neamar/kiss/IconsHandler.java:
```diff
@@ -0,0 +1,283 @@
+                    eventType = xpp.next();
+                }
+            }
+        } catch (NameNotFoundException e) {
```
```
}
catch
```
Same

Maybe do something about these catch?

saveman71 commented:
Some code style issues, but not critical at all! (I am not a Java developer, nor an Android one, so I'll leave to @Neamar the task to review the correctness of the code itself).

Obviously code style issues were reported once, but apply to the entire PR :)

ilmich commented:
Second iteration ;) 
I've fixed code style, some little adjustement and I've implemented a simple cache for generated icons. 

Size of cache depends:

1) size of original icons ( low screen density ~= low icon resolution)
2) number of missed icons 

Cheers
Michele

Neamar commented:
Hi @ilmich,

Sorry for the delay, I've been busy lately.... looks like there is some merge conflict, do you mind fixing them so I can put everything into master? Thanks!

ilmich commented:
Hi @Neamar, I'll fix this soon.
Sorry but I was on vacation last week :)

Neamar commented:
No problem :)

To be honest, I try to review the PR but it's way out of my league,  can't understand much... so I think once you've merged master, I'll probably merge without further comments ;)

Just one thing: maybe we can update the strings to indicate to the user which themes are supported, so they know what to look for in the Store? Or is any theme supported?

ilmich commented:
Hi @Neamar, there is something like a standard for developing icons pack if you want to try to understand better what my code does

http://forum.xda-developers.com/showthread.php?t=2399426

and also some empty templates

https://github.com/the1dynasty/IconPack-Template/tree/master/IconPack-Template/res

Regarding theme support, my code scan for all ADW.Launcher installed themes (look `loadAvailableIconsPacks()` method), that seems to be one of the most supported launcher.
But I've also defined an ad-hoc intent filter string `fr.neamar.kiss.THEMES` in order, if you want, to start to ask icons pack developers to support directly KISS Launcher.

Michele

ThePreviousOne commented:
Not to be rude but, I have an Icon Theme installed and it works, always has. Why make it more complicated? (I imagine I'm just misunderstanding the purpose of this PR, but this is the way it looks to me)

ilmich commented:
Hi @ThePreviousOne, I suppose that you have a Cyanogen based ROM or similar with a theme engine... but with pure android roms this is the only way to customize icons.

ThePreviousOne commented:
@ilmich Sorry, ihavent used stock in years, kind forgot the differences

licaon-kter commented:
Any news on this?

pcomans commented:
Bump! Any news on this @Neamar?

Neamar commented:
Hey @ilmich! Sorry for letting this one alone for so long...

But I'm (kinda) back!

I tested the PR with this theme: https://play.google.com/store/apps/details?id=com.nuyork.adwblack

In the settings:
![2016-02-29 21 34 19](https://cloud.githubusercontent.com/assets/536844/13409785/5f99d930-df2c-11e5-8a0b-b57d89ebdaf9.png)

But then:
![2016-02-29 21 34 12](https://cloud.githubusercontent.com/assets/536844/13409780/5d18583a-df2c-11e5-9c2c-3c7f365dc910.png)

Am I doing something wrong?

Thanks!

licaon-kter commented:
Restarting KISS does not help?

Neamar commented:
Nope :(

ilmich commented:
Hi @Neamar ... can you please try with another, and less older icons pack?!?!
For example 

https://play.google.com/store/apps/details?id=nexbit.icons.moonshine
https://play.google.com/store/apps/details?id=com.numix.icons_circle
https://play.google.com/store/apps/details?id=com.xda.miroex

Thanks

Neamar commented:
It worked :)

Neamar commented:
After a very long wait, welcome to `master` ;)

ilmich commented:
Thaks @Neamar  :)

saveman71 commented:
Hey @ilmich,

I'm currently working on your code, because your icon cache seems broken. For example, at line https://github.com/Neamar/KISS/pull/327/files#diff-3891f346f3ed7aab0a4e211c7193066eR197 you don't check that icon may be cached before calling `pm.getActivityIcon(componentName);`.

I implemented an other solution that uses the data structure below to sotre the drawables:

```
    // map with cached icon drawables (maxed for memory optimization)
    private LinkedHashMap<String, Drawable> iconCache = new LinkedHashMap<String, Drawable>() {
        @Override protected boolean removeEldestEntry(Entry<String, Drawable> entry) {
            return size() > 64; // arbitrary number here, can be changed without harm
        }
    };
```

This stores up to 64 different `Drawables` before dropping the older ones on insert. The advantage is that you can use the standard library calls to check if a drawable is in cache, retrieve a drawable, etc.

Was there a specific reason to use a fs-based cache system?

ilmich commented:
Hi @saveman71, this simple cache implementation is intended only for generated icons (that can be a more intensive task than simply get a drawable with android system calls).
I've decided to implemented a filesystem cache because in my mind this kind of cache should be persistent across reboot of phone or kiss restart.

But feel free to discuss more in depth another solution :)

thanks.

Neamar commented:
@ilmich I was reading http://openopensource.org/ and I felt that it's the right spirit, would you be interested to be made a collaborator on KISS? I think you deserve it with your contributions, and you can bring value to the project :)

ilmich commented:
@Neamar thanks for your words and I would be truly honored :)

Cheers
Michele

Neamar commented:
Done :)

Can you send me your email at neamar@neamar.fr? Thanks!

On Fri, Dec 23, 2016 at 7:12 AM Michele Zuccalà <notifications@github.com>
wrote:

> @Neamar <https://github.com/Neamar> thanks for your words and I would be
> truly honored :)
>
> Cheers
> Michele
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Neamar/KISS/pull/327#issuecomment-268945277>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAgxDKzS8yrojNhf2Pq3xkkkwyt6MgOUks5rK2ZHgaJpZM4Gm55g>
> .
>
